 Gerekli Kütüphaneler

In [ ]:
pip install pandas mlxtend


Apriori Algoritması ile Birliktelik Kuralları

In [ ]:
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

# Örnek veriseti: Alışveriş sepeti verisi
dataset = [
    ['süt', 'ekmek', 'yağ'],
    ['ekmek', 'yağ'],
    ['süt', 'ekmek'],
    ['süt', 'yağ'],
    ['ekmek', 'yağ']
]

# Veriyi one-hot encode et
from mlxtend.preprocessing import TransactionEncoder
te = TransactionEncoder()
te_ary = te.fit(dataset).transform(dataset)
df = pd.DataFrame(te_ary, columns=te.columns_)

# Sık öğe kümelerini bul
frequent_itemsets = apriori(df, min_support=0.4, use_colnames=True)

# Birliktelik kurallarını oluştur
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.7)

print("Sık Öğeler:")
print(frequent_itemsets)
print("\nBirliktelik Kuralları:")
print(rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']])


Sık Öğeler:
   support      itemsets
0      0.8       (ekmek)
1      0.6         (süt)
2      0.8         (yağ)
3      0.4  (süt, ekmek)
4      0.6  (ekmek, yağ)
5      0.4    (süt, yağ)

Birliktelik Kuralları:
  antecedents consequents  support  confidence    lift
0     (ekmek)       (yağ)      0.6        0.75  0.9375
1       (yağ)     (ekmek)      0.6        0.75  0.9375


Açıklamalar
TransactionEncoder: Liste formatındaki veriyi 0/1 tablosuna dönüştürür.

apriori: Belirli bir minimum destek eşiğine göre sık öğe kümelerini bulur.

association_rules: Bu kümelerden güven ve lift gibi ölçütlerle kurallar üretir.

Kısaca yorumlamak gerekirse:

Ekmek ⇒ Yağ ve Yağ ⇒ Ekmek kuralları, %75 güven ile doğru ama güçlü bir ilişki değil.

Lift değeri 0.9375, yani bu ürünler birlikte alınsa da aralarındaki ilişki rastlantısal olabilir.

Sonuç: Ekmek ve yağ sıklıkla birlikte alınsa da, biri diğerini alma ihtimalini çok artırmıyor.